engine/lite/runtime/src/main/scala/pl/touk/nussknacker/engine/lite/TestRunner.scala

```diff
@@ -39,11 +50,27 @@ class InterpreterTestRunner[F[_] : InterpreterShape : CapabilityTransformer : Ef
 
     //TODO: probably we don't need statics here, we don't serialize stuff like in Flink
     val collectingListener = ResultsCollectingListenerHolder.registerRun(variableEncoder)
-
     //in tests we don't send metrics anywhere
     val testContext = LiteEngineRuntimeContextPreparer.noOp.prepare(testJobData(process))
     val componentUseCase: ComponentUseCase = ComponentUseCase.TestRuntime
 
+    val expressionEvaluator: (Expression, Parameter, Context, NodeId) => Validated[NonEmptyList[PartSubGraphCompilationError], (String, AnyRef)] = {
+      val validationContext = GlobalVariablesPreparer(modelData.processWithObjectsDefinition.expressionConfig).emptyValidationContext(process.metaData)
+      val evaluator = ExpressionEvaluator.unOptimizedEvaluator(modelData)
+      val expressionCompiler = ExpressionCompiler.withoutOptimization(modelData).withExpressionParsers {
+        case spel: SpelExpressionParser => spel.typingDictLabels
+      }
+      (expression: Expression, parameter: Parameter, context: Context, nodeId: NodeId) => {
+        expressionCompiler
+          .compile(expression, Some(parameter.name), validationContext, parameter.typ)(nodeId)
```

Comment:
We still have some problem here. The correct solution would be that you did this compilation before `prepareRecordForTest` and instead of passing `Source`, pass compiled parameters (some `F[List[TypedExpression]]`) Thanks to that we would pass simple method without such many parameters. If you don't have energy to fix that, you can leave here only a TODO: compile parameters once

Reply:
Hmm, think I know what you mean but I am missing `testParametersDefinition` from source to get list of parameters which I need to execute validation. I could do that by moving logic that match source for `TestWithParametersSupport` a little bit earlier but it still doesn't solve problem you pointed out :thinking:

Reply:
I'm wrong, you're right. It is impossible to compile expression once because it will be different for each record - it is other case than in production code parameters handling